crates/core/component/dex/src/component/position_manager.rs

```diff
@@ -455,5 +460,139 @@ pub(super) trait Inner: StateWrite {
+                });
+                let old_a = Balance::from(Value {
+                    asset_id: pair.asset_1,
+                    amount: prev
+                        .reserves_for(pair.asset_1)
+                        .expect("specified position should match provided trading pair"),
+                });
+                let old_b = Balance::from(Value {
+                    asset_id: pair.asset_2,
+                    amount: prev
+                        .reserves_for(pair.asset_2)
+                        .expect("specified position should match provided trading pair"),
+                });
+
+                (new_a - old_a, new_b - old_b)
+            }
+            (State::Closed, Some(prev)) => {
+                // The previous amount of assets A and B should be subtracted from the aggregate value.
+
+                let pair = position.phi.pair;
+                let old_a = prev
+                    .reserves_for(pair.asset_1)
+                    .expect("specified position should match provided trading pair");
+                let old_b = prev
+                    .reserves_for(pair.asset_2)
+                    .expect("specified position should match provided trading pair");
+
+                let old_a = Balance::from(Value {
+                    asset_id: pair.asset_1,
+                    amount: old_a,
+                });
+                let old_b = Balance::from(Value {
+                    asset_id: pair.asset_2,
+                    amount: old_b,
+                });
+                // The position is closed, so the delta is the negative of the previous reserves.
+                (-old_a, -old_b)
+            }
+            (State::Withdrawn, _) | (State::Claimed, _) | (State::Closed, None) => {
+                // The position already went through the `Closed` state or was opened in the `Closed` state, so its contribution has already been subtracted.
+                return Ok(());
+            }
+        };
+
+        tracing::debug!(
+            ?position,
+            ?delta_a,
+            ?delta_b,
+            "updating position assets' aggregate balances"
+        );
+
+        let mut value_circuit_breaker: ValueCircuitBreaker = match self
+            .nonverifiable_get_raw(&state_key::aggregate_value().as_bytes())
+            .await
+            .expect("able to retrieve value circuit breaker from nonverifiable storage")
+        {
+            Some(bytes) => bincode::deserialize(&bytes).expect(
+                "able to deserialize stored value circuit breaker from nonverifiable storage",
+            ),
+            None => ValueCircuitBreaker::default(),
+        };
+
+        // Add the delta to the value circuit breaker for assets A and B.
+        value_circuit_breaker.tally(delta_a);
+        value_circuit_breaker.tally(delta_b);
+
+        // Confirm that the value circuit breaker is still within the limits.
+        value_circuit_breaker.check()?;
```

Comment:
It's late so maybe I'm missing something, but my first read of this PR is that the balances are aggregated correctly, but the circuit breaker check happens at the wrong resolution. We do it every time we update a position, but that means we can't factor in the "outflow" (the swap output) and so it will detect benign state updates at value breaking.

If I try to run through an example:
Let's take a position $P$ with $p,q=1$, and reserves $(R_1, R_2) = (100, 0)$, suppose that the position is updated and "flips" completely on the other side such that we have $(R_1^*, R_2^*) = (0, 100)$.

By applying the above logic, we compute the aggregated balance of the value circuit breaker (`(new_a - old_a, new_b - old_b)`):

$T_A = 0 - 100$ of asset 1

$T_B = 100 - 0$ of asset 2.

So the circuit breaker check will fail.

Reply:
I could be missing something, but it seems to me we want to perform checks at two levels:
- when we update a position
- when we collect the batched swap outputs from the dex engine

Reply:
> By applying the above logic, we compute the aggregated balance of the value circuit breaker ((new_a - old_a, new_b - old_b)):

I don't think this matches how it's implemented, unless I am misunderstanding.

What I believe is implemented (which should be validated with integration tests) is that:

* When the position is initially stored with reserves `(100, 0)`, `delta_a = +100 asset_1` and `delta_b = 0 asset_2`. 
* These are tallied to the circuit breaker and thus it has balances of `asset_1: 100, asset_2: 0`
* When the position is filled against and the reserves flip, the `(new_a - old_a, new_b - old_b)` are assigned to `(delta_a, delta_b)`. `delta_a = -100 asset_1` and `delta_b = +100 asset_2`, and then we add `delta_a` to the aggregate balance (by calling `value_circuit_breaker.tally(delta_a)`) which should result in a balance of `asset_1: 0, asset_2: 100` being tracked

Reply:
Ah I see, that makes total sense

Reply:
As for where we perform checks, I _think_ this is okay -- at the time the batch swap outputs are minted, the positions that were filled against (should) have already all been updated to reflect the in/outflows. I suppose this could miss a case where `put_position` was not called on positions that were filled against, but I don't think it should be possible given the way the dex is implemented (unless we fail to store the updated positions to storage or something??)

Reply:
I think we can add one additional check to `handle_batch_swaps`:

* At the beginning of the method, fetch the aggregate balances for asset_1/2 prior to calling `route_and_fill`
* At the end of the method perform a check that the outflows do not exceed the aggregate balance for either asset

Reply:
That sounds good!